lib/ably/rest/channel.rb

```diff
@@ -74,7 +78,12 @@ def publish(name, data = nil, attributes = {})
             unless client.auth.can_assume_client_id?(msg.client_id)
               raise Ably::Exceptions::IncompatibleClientId.new("Cannot publish with client_id '#{msg.client_id}' as it is incompatible with the current configured client_id '#{client.client_id}'")
             end
-          end.as_json
+          end.as_json.tap do |msg|
+            # Mutate the JSON to support idempotent publishing where a Message.id does not exist
+            unless msg['id']
+              msg['id'] = SecureRandom.hex(16).force_encoding('UTF-8')
+            end
```

Comment:
so shouldn't this be conditional on `if client.idempotent_rest_publishing`?

Reply:
also, why not save some bandwidth and use base64 rather than hex, 50% more entropy per character

Reply:
> so shouldn't this be conditional on client.idempotent_rest_publishing?

Yes.  This change did in fact break some tests, which I never got around to checking.  Good spot, but thankfully the tests broke too as expected!

I've also implemented the `randomId:0` functionality but it gets rejected by realtime:

```
19:49:24.877 [debug] => URL: post https://local-rest.ably.io/channels/e60f1f1eb9b009f5f5399c1c1ce5f950/publish, Headers: Content-Type: "application/json", Accept: "application/json", User-Agent: "Ably Ruby client 1.0.7 (https://www.ably.io)", X-Ably-Version: "1.0", X-Ably-Lib: "ruby-1.0.7", Authorization: "Basic X3RtcF9lRVdWcFEuYzMwUFp3Omtab09MQnpKWXZfUUEyRk0="
19:49:24.878 [debug] => Body: [{"data":"7724cf521aab5a3bd8fd1c18229948b8","id":"is2mDOHSn30fUDFhY/Bnxw==:0"},{"data":"7724cf521aab5a3bd8fd1c18229948b8","id":"is2mDOHSn30fUDFhY/Bnxw==:1"},{"data":"7724cf521aab5a3bd8fd1c18229948b8","id":"is2mDOHSn30fUDFhY/Bnxw==:2"}]
19:49:24.918 [debug] <= Status: 400, Headers: X-Ably-ErrorCode: "40007", X-Ably-ErrorMessage: "Invalid request: multiple messages in a single request with a client-specified id", Content-Type: "application/json", Content-Length: "153", Access-Control-Expose-Headers: "Link,Transfer-Encoding,Content-Length,X-Ably-ErrorCode,X-Ably-ErrorMessage,X-Ably-ServerId,Server", Access-Control-Allow-Origin: "*", Access-Control-Allow-Credentials: "true", Vary: "Origin", Date: "Tue, 24 Jul 2018 18:49:24 GMT", Connection: "keep-alive", Ably-Middleware-Parsed: true
19:49:24.918 [debug] <= Body: {"error"=>{"statusCode"=>400, "code"=>40007, "message"=>"Invalid request: multiple messages in a single request with a client-specified id"}}
```

@SimonWoolf @paddybyers what is wrong with that publish request?

Reply:
> what is wrong with that publish request?

> Invalid request: multiple messages in a single request with a client-specified id

According to what's currently implemented based on my proposal (https://github.com/ably/wiki/issues/191#issuecomment-397741807) we only support a single message in a REST publish if the client supplies the `id`.

The final spec is up for discussion.